src/NuGet.Services.AzureSearch/Owners2AzureSearch/OwnerSetComparer.cs

```diff
@@ -0,0 +1,91 @@
+        }
+
+        public SortedDictionary<string, string[]> Compare(
+            SortedDictionary<string, SortedSet<string>> oldData,
+            SortedDictionary<string, SortedSet<string>> newData)
+        {
+            // We use a very simplistic algorithm here. Perform one pass on the new data to find the added or changed
+            // package IDs. Then perform a second pass on the old data to find removed package IDs. We can optimize
+            // this later if necessary.
+            //
+            // We emit all of the usernames when a package ID's owners have changed instead of the delta. This is
+            // because Azure Search does not have a way to append or remove a specific item from a field that is an
+            // array. The entire new array needs to be provided.
+            var result = new SortedDictionary<string, string[]>(StringComparer.OrdinalIgnoreCase);
+
+            // First pass: find added or changed sets.
+            foreach (var pair in newData)
+            {
+                var id = pair.Key;
+                var newOwners = pair.Value;
+                if (!oldData.TryGetValue(id, out var oldOwners))
+                {
+                    // ADDED: The package ID does not exist in the old data, which means the package ID was added.
+                    result.Add(id, newOwners.ToArray());
+                    _logger.LogInformation(
+                        "The package ID {ID} has been added, with {AddedCount} owners.",
+                        id,
+                        newOwners.Count);
+                }
+                else
+                {
+                    // The package ID exists in the old data. We need to check if the owner set has changed. Perform
+                    // an ordinal comparison to allow username case changes to flow through.
+                    var removedUsernames = oldOwners.Except(newOwners, StringComparer.Ordinal).ToList();
+                    var addedUsernames = newOwners.Except(oldOwners, StringComparer.Ordinal).ToList();
+
+                    if (removedUsernames.Any() || addedUsernames.Any())
+                    {
+                        // CHANGED: The username set has changed.
+                        result.Add(id, newOwners.ToArray());
+                        _logger.LogInformation(
+                            "The package ID {ID} has an ownership change, with {RemovedCount} owners removed and " +
+                            "{AddedCount} owners added.",
+                            id,
+                            removedUsernames.Count,
+                            addedUsernames.Count);
+                    }
+                }
+            }
+
+            // Second pass: find removed sets.
+            foreach (var pair in oldData)
+            {
+                var id = pair.Key;
+                var oldOwners = pair.Value;
+
+                if (!newData.TryGetValue(id, out var newOwners))
```

Comment:
Assuming that the two sets are mostly the same, we're doing a comparison here that we've _usually_ already done above. Is there a way we could filter out the IDs we've already processed from this second pass, or rewrite the algorithm so that we only process each ID once? This could become very expensive if we're comparing millions of packages owners at once.

For example, provided that `Distinct` isn't very expensive, potentially something like this could be much faster.

```
var ids = oldData.Keys.Concat(newData.Keys).Distinct();
foreach (var id in ids)
{
    var wasInOld = oldData.TryGetValue(id, out var oldOwners);
    var isInNew = newData.TryGetValue(id, out var newOwners);
    if (wasInOld)
    {
        if (isInNew)
        {
            // check for changes
        }
        else
        {
            // removed
        }
    }
    else
    {
        if (isInNew)
        {
            // added
        }
        else
        {
            // throw exception because this shouldn't happen
        }
    }
}
```

Reply:
I am not exactly sure how `Distinct` is implemented by my guess is that it uses a hash set internally. This means that in the case where `oldData` and `newData` are the same (the extremem closest to what is going to happen the most), we'll be performing 4 * N hash look-ups. 2 * N in `Distinct` and 2 * N in the `foreach`. This should become obvious with some basic performance testing. I'll try to find some time today to do this.

Reply:
DEV data, 100 iterations, no debugger attached, my devbox. Averaged values.

Existing algorithm in PR: 381ms.
Suggested algorithm: 427ms.

In short, `Distinct` is not magic. I looked at the implementation. It uses a hash set.